src/builder.c

```diff
@@ -113,6 +113,53 @@ int trilogy_builder_write_uint64(trilogy_builder_t *builder, uint64_t val)
     return TRILOGY_OK;
 }
 
+#ifndef TRILOGY_FLOAT_BUF
+#define TRILOGY_FLOAT_BUF
+
+typedef union {
+    float f;
+    uint32_t u;
+} trilogy_float_buf_t;
+
+typedef union {
+    double d;
+    uint64_t u;
+} trilogy_double_buf_t;
+
+#endif
```

Comment:
Should the block above instead be in a header? If it's not, I don't think the define guards would be needed since `*.c` files should only be compiled once?

Reply:
To be honest, I kinda hated doing this 😅

But, these types are truly just helpers for the internal serialization implementation(s) and should never have to be used by a public user of this library. I know it's not a big deal to throw it in a header and just mark it private, but as we've both seen with `libmysqlclient` - whatever is exposed to the caller, they'll use and hack in ways we can't imagine.

If this stays in the C files, the guards are because the Ruby extension combines all the C files into a single source file for compilation. That process was actually something I wanted to have as a `Makefile` target for folks to more easily be able to embed this project in their own (which would combine everything into a single `.c`. and `.h`). I haven't checked if it exists yet, but it would be nice if it did.

That said, I could pretty easily be convinced to just put it in a header.

What do you think?

Reply:
I think the guards were related to the Ruby library concatenating all the source files: https://github.com/github/trilogy/blob/main/contrib/ruby/ext/trilogy-ruby/extconf.rb#L3-L6

Reply:
Ha! You beat me to this comment by a minute

Reply:
@brianmario Would it alternatively work to do this inline? For example:

```c
int trilogy_reader_get_double(trilogy_reader_t *reader, double *out)
    union {double d; uint64_t u;} double_val;
    double_val.d = val;

    int rc = trilogy_reader_get_uint64(reader, &double_val.u);
    if (rc != TRILOGY_OK) {
        return rc;
    }

    *out = double_val.d;
    return TRILOGY_OK;
}
```

That way there's no separate type definition necessary and there's no issue with combining C files either?

Reply:
Ah yeah I like that 👍